File: scripts/verify-chain.js
const Database = require('better-sqlite3');
const crypto = require('crypto');
const path = require('path');

const DB_PATH = process.argv[2] || path.join(process.env.MBO_PROJECT_ROOT || process.cwd(), '.mbo', 'mirrorbox.db');

/**
 * Section 7: Audit Tool
 * Verifies Invariant 4: Every operation is reproducible from the event store.
 * Checks SHA-256 hashes, sequence continuity, and hash chaining (prev_hash).
 * Scaled for multi-million row databases via cursor iteration.
 */
function verifyChain() {
  console.log(`Auditing Mirror Box Chain of Custody: ${DB_PATH}`);
  const db = new Database(DB_PATH, { readonly: true });

  try {
    const statement = db.prepare('SELECT id, seq, stage, actor, timestamp, payload, hash, world_id, parent_event_id, prev_hash FROM events ORDER BY seq ASC');

    let lastId = null;
    let expectedSeq = 1;
    let errors = 0;
    let maxSeq = 0;
    let lastHash = null;
    let count = 0;

    for (const event of statement.iterate()) {
      // 1. Verify Sequence Continuity
      if (event.seq !== expectedSeq) {
        console.error(`\nFAIL [Sequence Gap]: Expected seq ${expectedSeq}, found ${event.seq} (Event ${event.id})`);
        errors++;
      }
      expectedSeq++;
      maxSeq = event.seq;

      // 2. Verify Hash Chaining
      if (count === 0) {
        if (event.prev_hash !== null) {
          console.error(`\nFAIL [Root Chain]: Root event ${event.id} should have prev_hash null, found ${event.prev_hash}`);
          errors++;
        }
      } else {
        if (event.prev_hash !== lastHash) {
          console.error(`\nFAIL [Chain Breach]: Event ${event.id} (seq ${event.seq}) refers to prev_hash ${event.prev_hash}, expected ${lastHash}`);
          errors++;
        }
      }

      // 3. Verify Hash Integrity
      const envelope = JSON.stringify({
        id: event.id,
        seq: event.seq,
        stage: event.stage,
        actor: event.actor,
        timestamp: event.timestamp,
        world_id: event.world_id,
        parent_event_id: event.parent_event_id ?? null,
        prev_hash: event.prev_hash ?? null,
        payload: event.payload
      });
      const calculatedHash = crypto.createHash('sha256').update(envelope).digest('hex');
      if (calculatedHash !== event.hash) {
        console.error(`\nFAIL [Hash Mismatch]: Event ${event.id} (seq ${event.seq})`);
        errors++;
      }

      // 4. Verify Parent Linkage
      if (count > 0 && event.parent_event_id !== lastId) {
        console.error(`\nFAIL [Broken Link]: Event ${event.id} refers to parent ${event.parent_event_id}, expected ${lastId}`);
        errors++;
      }

      lastId = event.id;
      lastHash = event.hash;
      count++;

      if (count % 50000 === 0) {
        process.stderr.write(`\rVerified ${count} events...`);
      }
    }

    process.stderr.write(`\rVerification complete. Checking anchor...\n`);

    if (count === 0) {
      console.log('PASS: Event store is empty.');
      return;
    }

    // 5. Verify Anchor — uses run_id PK (migrated schema)
    const anchor = db.prepare('SELECT run_id, seq, event_id, hash FROM chain_anchors ORDER BY created_at DESC LIMIT 1').get();
    if (!anchor) {
      console.error(`FAIL [Missing Anchor]: No anchor found in chain_anchors.`);
      errors++;
    } else if (anchor.hash !== lastHash) {
      console.error(`FAIL [Anchor Mismatch]: Anchor hash ${anchor.hash.slice(0,12)} does not match chain tail hash ${lastHash.slice(0,12)}.`);
      errors++;
    }

    if (errors === 0) {
      console.log(`PASS: Verified ${count} events. Chain intact from seq 1 → ${maxSeq}.`);
      console.log(`Anchor run_id: ${anchor.run_id} | tail hash: ${lastHash.slice(0,16)}...`);
    } else {
      console.error(`FAILED: ${errors} integrity error(s) found.`);
      process.exit(1);
    }
  } catch (error) {
    console.error(`\nERROR: ${error.message}`);
    process.exit(1);
  } finally {
    db.close();
  }
}

verifyChain();

File: scripts/verify-chain.py
#!/usr/bin/env python3
import sqlite3
import json
import hashlib
import sys
import os
import platform
import getpass

# ── Rule 1: Environment Assertions ───────────────────────────────────────────
ENV_CHECKS = [
    {"label": "Platform is macOS", "ok": platform.system() == "Darwin", "detail": f"got {platform.system()}"},
    {"label": "Running as johnserious", "ok": getpass.getuser() == "johnserious", "detail": f"got {getpass.getuser()}"}
]

env_failed = False
print("── Environment checks ──────────────────────────────────────────")
for check in ENV_CHECKS:
    status = "PASS" if check["ok"] else "FAIL"
    detail = f" — {check['detail']}" if not check["ok"] else ""
    print(f"{status}  {check['label']}{detail}")
    if not check["ok"]:
        env_failed = True

if env_failed:
    print("\nEnvironment checks failed — aborting.\n")
    sys.exit(2)
print("── Environment OK — proceeding with chain verification ───────────\n")

DEFAULT_DB_CANDIDATES = [
    os.path.join(os.path.dirname(__file__), '../.mbo/mirrorbox.db'),
    os.path.join(os.path.dirname(__file__), '../data/mirrorbox.db'),
]

DB_PATH = sys.argv[1] if len(sys.argv) > 1 else next((p for p in DEFAULT_DB_CANDIDATES if os.path.exists(p)), DEFAULT_DB_CANDIDATES[0])

results = []
def record(label, ok):
    results.append({"label": label, "status": "PASS" if ok else "FAIL"})
    if ok:
        print(f"  ✓ {label}")
    else:
        print(f"  ✗ {label}")

def verify_chain():
    print(f"Auditing Mirror Box Chain of Custody: {DB_PATH}\n")

    if not os.path.exists(DB_PATH):
        print(f"ERROR: Database not found at {DB_PATH}")
        sys.exit(1)

    conn = None
    try:
        conn = sqlite3.connect(f"file:{DB_PATH}?mode=ro", uri=True)
        cursor = conn.cursor()

        cursor.execute("SELECT id, seq, stage, actor, timestamp, payload, hash, world_id, parent_event_id, prev_hash FROM events ORDER BY seq ASC")
        events = cursor.fetchall()

        if not events:
            record("Event store is empty (Valid State)", True)
            return

        expected_seq = 1
        errors = 0
        last_hash = None
        last_id = None

        for index, event in enumerate(events):
            e_id, e_seq, e_stage, e_actor, e_timestamp, e_payload, e_hash, e_world_id, e_parent_id, e_prev_hash = event

            # 1. Sequence continuity
            if e_seq != expected_seq:
                errors += 1
            expected_seq += 1

            # 2. Hash chaining
            if index > 0 and e_prev_hash != last_hash:
                errors += 1

            # 3. Hash integrity
            envelope = json.dumps({
                "id": e_id,
                "seq": e_seq,
                "stage": e_stage,
                "actor": e_actor,
                "timestamp": e_timestamp,
                "world_id": e_world_id,
                "parent_event_id": e_parent_id,
                "prev_hash": e_prev_hash,
                "payload": e_payload
            }, separators=(',', ':'), ensure_ascii=False)

            calculated_hash = hashlib.sha256(envelope.encode('utf-8')).hexdigest()
            if calculated_hash != e_hash:
                errors += 1

            last_id = e_id
            last_hash = e_hash

        record(f"Verified {len(events)} events (Hash Integrity)", errors == 0)

        # 5. Anchor verification
        cursor.execute("SELECT run_id, seq, event_id, hash FROM chain_anchors ORDER BY created_at DESC LIMIT 1")
        anchor = cursor.fetchone()
        anchor_ok = False
        if anchor:
            a_run_id, a_seq, a_event_id, a_hash = anchor
            anchor_ok = (a_hash == last_hash)
        
        record("Chain Anchor Alignment", anchor_ok)

        # ── Rule 5: Live Output Table ──────────────────────────────────────────
        print("\n── Chain Audit Summary Table ────────────────────────────────────")
        print("| Audit Check                                    | Status       |")
        print("| ---------------------------------------------- | ------------ |")
        for res in results:
            print(f"| {res['label']:<46} | {res['status']:<12} |")
        print("────────────────────────────────────────────────────────────────\n")

        if errors > 0 or not anchor_ok:
            sys.exit(1)

    except sqlite3.Error as e:
        print(f"ERROR: {e}")
        sys.exit(1)
    finally:
        if conn:
            conn.close()

if __name__ == "__main__":
    verify_chain()
